### PR TITLE
[ The final addition for complete and smooth integration ]Fix minor stuff + refine flow + change some data types for smoother integration + add lecture model and populated it in get CoursebyId for better integration

### PR DESCRIPTION
**Pull request by Devesh21700Kumar**
**Description**

** HOT FIX **

Fix minor stuff + refine flow + change some data types for smoother integration + add lecture model and populated it in get CoursebyId for better integration

Added a new lecture schema with documentation and fixed tests.
Added its endpoints to suit the workflow of course -> schedule -> lecture

**Type of Change:**

- [x] Code
- [x] New Feature
- [x] Documentation
- [x] Testing

**How Has This Been Tested?**

Terminal commands to test code:-
1). npm run lint
2). npm run lint:fix
3). nest start
4). npm run test

**Checklist**

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] My changes generate no new warnings
- [x] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

**Review by Devesh21700Kumar — COMMENTED** [review]
@Abhishek-kumar09  In th emeet today, a lot of  integration was finalised and some extra fields had to be added with the following workflow that will help smoothen the process

File: src/modules/user/user.service.ts [hunk 1]
```diff
@@ -59,7 +59,8 @@ export class UserService {
       const { email, fId, role } = request['user'];
       const userExists = await this.userModel.findOne({ email: email }).lean();
       if (userExists) {
-        throw new ConflictException('User already exists');
+        console.log('User Already Exists');
+        return;
```
Comment (Abhishek-kumar09) [inline src/modules/user/user.service.ts:63]: change this

Comment (Devesh21700Kumar) [reply src/modules/user/user.service.ts:63]: The issue has been solved by creating a user if it does not exist in the middleware. This solves the problem of login, logout and relogin and also prevents duplicacy by caching and this flow is also used in most of the Mern apps using firebase authentication

**Review by Devesh21700Kumar — COMMENTED** [review]
Added isUpcoming field in course shcema for 
isUpcoming implementation

File: src/middleware/preAuth.middleware.ts [hunk 2]
```diff
@@ -27,6 +27,12 @@ export class PreauthMiddleware implements NestMiddleware {
           if (userExists) {
             role = userExists.role || Role.STUDENT;
           } else {
+            const newUser = new this.userModel({
+              email,
+              fid: uid,
+              role: Role.STUDENT,
+            });
+            newUser.save();
```
Comment (Abhishek-kumar09) [inline src/middleware/preAuth.middleware.ts:35]: you will also get photo and display name from the firebase. You can add that too.

Comment (Devesh21700Kumar) [reply src/middleware/preAuth.middleware.ts]: Sure !
I added photo url now but as per latest documentation admin.auth().verifyIdToken() no longer returns the name of the user

But the client-side can get it from localstorage as it is present there in the object and use it

File: src/middleware/preAuth.middleware.ts [hunk 3]
```diff
@@ -5,6 +5,7 @@ import { NextFunction, Request, Response } from 'express';
 import admin from '../main';
 import { Model } from 'mongoose';
 import { Role } from '../roles/role.enum';
+import { PictureAsPdf } from '@material-ui/icons';
```
Comment (Abhishek-kumar09) [inline src/middleware/preAuth.middleware.ts:8]: remove this

File: src/middleware/preAuth.middleware.ts [hunk 4]
```diff
@@ -16,7 +17,7 @@ export class PreauthMiddleware implements NestMiddleware {
         .auth()
         .verifyIdToken(token.replace('Bearer ', ''))
         .then(async (decodedToken) => {
-          const { email, uid } = decodedToken;
+          const { email, uid, picture } = decodedToken;
```
Comment (Abhishek-kumar09) [inline src/middleware/preAuth.middleware.ts:20]: add display name also.
And store it in firstName field with spliting through space.

File: src/middleware/preAuth.middleware.ts [hunk 5]
```diff
@@ -16,7 +16,7 @@ export class PreauthMiddleware implements NestMiddleware {
...
```
Comment (Abhishek-kumar09) [inline src/middleware/preAuth.middleware.ts:19]: I don't think picture is any key in decodedToken.
Please correct it.
It is photoUrl

Comment (Abhishek-kumar09) [reply src/middleware/preAuth.middleware.ts:19]: If I am correct.

Comment (Devesh21700Kumar) [reply src/middleware/preAuth.middleware.ts:19]: Oh. it has the picture property
photoUrl is not there , just checked it again

**Pull request merged**